File: secure-gateway-studio/extension/src/domain/iam-policy.ts
import { canonicalJson } from "./canonical.ts";

export interface IamBinding {
  role?: unknown;
  members?: unknown;
  condition?: unknown;
  [key: string]: unknown;
}

export interface IamPolicy {
  bindings?: unknown;
  version?: unknown;
  etag?: unknown;
  auditConfigs?: unknown;
  [key: string]: unknown;
}

function sameCondition(left: unknown, right: unknown): boolean {
  return canonicalJson(left ?? null) === canonicalJson(right ?? null);
}

interface ParsedBinding {
  binding: IamBinding;
  key: string;
  members: string[];
}

function validateCondition(condition: unknown): void {
  if (condition === undefined) return;
  if (condition === null || typeof condition !== "object" || Array.isArray(condition)) {
    throw new Error("IAM binding contains an invalid condition");
  }
  const candidate = condition as Record<string, unknown>;
  const allowed = new Set(["title", "expression", "description", "location"]);
  if (
    Object.keys(candidate).some((field) => !allowed.has(field)) ||
    typeof candidate.title !== "string" || candidate.title.trim() === "" ||
    typeof candidate.expression !== "string" || candidate.expression.trim() === ""
  ) {
    throw new Error("IAM binding condition lacks a title or expression");
  }
  for (const optional of ["description", "location"] as const) {
    if (candidate[optional] !== undefined && typeof candidate[optional] !== "string") {
      throw new Error(`IAM binding condition contains an invalid ${optional}`);
    }
  }
}

function validateAuditConfigs(value: unknown): void {
  if (value === undefined) return;
  if (!Array.isArray(value)) throw new Error("IAM policy contains invalid auditConfigs");
  const services = new Set<string>();
  for (const raw of value) {
    if (raw === null || typeof raw !== "object" || Array.isArray(raw)) {
      throw new Error("IAM policy contains an invalid audit config");
    }
    const audit = raw as Record<string, unknown>;
    if (
      Object.keys(audit).some((field) => !["service", "auditLogConfigs"].includes(field)) ||
      typeof audit.service !== "string" || audit.service.trim() === "" ||
      services.has(audit.service) ||
      (audit.auditLogConfigs !== undefined && !Array.isArray(audit.auditLogConfigs))
    ) throw new Error("IAM policy contains an invalid audit config");
    services.add(audit.service);
    const logTypes = new Set<string>();
    for (const rawLog of (audit.auditLogConfigs ?? []) as unknown[]) {
      if (rawLog === null || typeof rawLog !== "object" || Array.isArray(rawLog)) {
        throw new Error("IAM policy contains an invalid audit log config");
      }
      const log = rawLog as Record<string, unknown>;
      if (
        Object.keys(log).some((field) => !["logType", "exemptedMembers"].includes(field)) ||
        typeof log.logType !== "string" ||
        !["ADMIN_READ", "DATA_WRITE", "DATA_READ"].includes(log.logType) ||
        logTypes.has(log.logType) ||
        (log.exemptedMembers !== undefined && !Array.isArray(log.exemptedMembers))
      ) throw new Error("IAM policy contains an invalid audit log config");
      logTypes.add(log.logType);
      if (Array.isArray(log.exemptedMembers)) {
        const members = new Set<string>();
        for (const member of log.exemptedMembers) {
          if (typeof member !== "string" || member.trim() === "" || members.has(member)) {
            throw new Error("IAM audit log config contains an invalid exempted member");
          }
          members.add(member);
        }
      }
    }
  }
}

function parseBindings(policy: IamPolicy): ParsedBinding[] {
  if (policy.bindings !== undefined && !Array.isArray(policy.bindings)) {
    throw new Error("IAM policy contains an invalid bindings collection");
  }
  const keys = new Set<string>();
  return ((policy.bindings ?? []) as unknown[]).map((raw) => {
    if (raw === null || typeof raw !== "object" || Array.isArray(raw)) {
      throw new Error("IAM policy contains an invalid binding");
    }
    const binding = structuredClone(raw as IamBinding);
    if (
      Object.keys(binding).some((field) => !["role", "members", "condition"].includes(field)) ||
      typeof binding.role !== "string" || binding.role.trim() === ""
    ) {
      throw new Error("IAM binding contains an invalid role");
    }
    if (!Array.isArray(binding.members) || binding.members.length === 0) {
      throw new Error("IAM binding contains an invalid members collection");
    }
    validateCondition(binding.condition);
    const seenMembers = new Set<string>();
    const members = binding.members.map((member) => {
      if (typeof member !== "string" || member.trim() === "") {
        throw new Error("IAM binding contains a non-string member");
      }
      if (seenMembers.has(member)) throw new Error("IAM binding contains duplicate members");
      seenMembers.add(member);
      return member;
    });
    const key = canonicalJson([binding.role, binding.condition ?? null]);
    if (keys.has(key)) throw new Error("IAM policy contains duplicate role/condition bindings");
    keys.add(key);
    return {
      binding,
      key,
      members,
    };
  });
}

/**
 * Validate every binding before a freshly read policy can be merged and sent
 * back to Google. Returning a clone prevents callers from accidentally
 * relying on unvalidated objects from the transport response.
 */
export function validatedIamPolicy(
  policy: IamPolicy,
  options: { requireEtag?: boolean } = { requireEtag: true },
): IamPolicy {
  if (policy === null || typeof policy !== "object" || Array.isArray(policy)) {
    throw new Error("IAM policy is not an object");
  }
  const allowed = new Set(["version", "bindings", "auditConfigs", "etag"]);
  if (Object.keys(policy).some((field) => !allowed.has(field))) {
    throw new Error("IAM policy contains unknown fields");
  }
  if (
    policy.version !== undefined &&
    (!Number.isInteger(policy.version) || ![0, 1, 3].includes(policy.version as number))
  ) throw new Error("IAM policy contains an invalid version");
  if (
    policy.etag !== undefined &&
    (typeof policy.etag !== "string" || policy.etag.trim() === "")
  ) throw new Error("IAM policy contains an invalid etag");
  if (
    options.requireEtag !== false &&
    (typeof policy.etag !== "string" || policy.etag.trim() === "")
  ) {
    throw new Error("IAM policy is missing an etag");
  }
  validateAuditConfigs(policy.auditConfigs);
  const bindings = parseBindings(policy);
  if (
    bindings.some(({ binding }) => binding.condition !== undefined) &&
    policy.version !== 3
  ) throw new Error("conditional IAM bindings require policy version 3");
  return {
    ...structuredClone(policy),
    bindings: bindings.map(({ binding }) => binding),
  };
}

/** Strict binding-only parser for durable bootstrap ownership pins. */
export function validatedIamBindings(value: unknown): IamBinding[] {
  const policy = validatedIamPolicy(
    { bindings: value },
    { requireEtag: false },
  );
  return structuredClone((policy.bindings ?? []) as IamBinding[]);
}

function memberGroups(bindings: ParsedBinding[]): Map<string, Set<string>> {
  const groups = new Map<string, Set<string>>();
  for (const binding of bindings) {
    const members = groups.get(binding.key) ?? new Set<string>();
    for (const member of binding.members) members.add(member);
    groups.set(binding.key, members);
  }
  return groups;
}

/**
 * Reverse only the role/condition/member delta produced by one IAM mutation.
 *
 * `beforePolicy` and `afterPolicy` are the exact policies checkpointed around
 * Apply. `currentPolicy` is a fresh v3 read at compensation time. Bindings and
 * members introduced by other administrators after Apply are therefore kept;
 * only members added by this run are removed and members removed by this run
 * are restored. Policy metadata and the fresh etag come from `currentPolicy`.
 */
export function revertIamPolicyDelta(options: {
  beforePolicy: IamPolicy;
  afterPolicy: IamPolicy;
  currentPolicy: IamPolicy;
}): IamPolicy {
  const beforePolicy = validatedIamPolicy(options.beforePolicy);
  const afterPolicy = validatedIamPolicy(options.afterPolicy);
  const currentPolicy = validatedIamPolicy(options.currentPolicy);
  const beforeBindings = parseBindings(beforePolicy);
  const afterBindings = parseBindings(afterPolicy);
  const currentBindings = parseBindings(currentPolicy);
  const beforeGroups = memberGroups(beforeBindings);
  const afterGroups = memberGroups(afterBindings);
  const affectedKeys = new Set([...beforeGroups.keys(), ...afterGroups.keys()]);
  let reverted = currentBindings.map(({ binding, key, members }) => ({
    binding,
    key,
    members: [...members],
  }));

  for (const key of affectedKeys) {
    const beforeMembers = beforeGroups.get(key) ?? new Set<string>();
    const afterMembers = afterGroups.get(key) ?? new Set<string>();
    const added = new Set([...afterMembers].filter((member) => !beforeMembers.has(member)));
    const removed = new Set([...beforeMembers].filter((member) => !afterMembers.has(member)));
    if (added.size === 0 && removed.size === 0) continue;

    let target = reverted.find((entry) => entry.key === key);
    for (const entry of reverted) {
      if (entry.key === key) {
        entry.members = entry.members.filter((member) => !added.has(member));
      }
    }
    if (removed.size > 0) {
      if (target === undefined) {
        const template =
          beforeBindings.find((entry) => entry.key === key) ??
          afterBindings.find((entry) => entry.key === key);
        if (template === undefined) {
          throw new Error("IAM delta lacks a binding template");
        }
        target = {
          binding: structuredClone(template.binding),
          key,
          members: [],
        };
        reverted.push(target);
      }
      target.members = [...new Set([...target.members, ...removed])].sort();
    }
    reverted = reverted.filter((entry) => entry.key !== key || entry.members.length > 0);
  }

  return validatedIamPolicy({
    ...structuredClone(currentPolicy),
    bindings: reverted.map(({ binding, members }) => ({ ...binding, members })),
    version: 3,
  });
}

/**
 * Replace only the members this run previously managed for one role/condition.
 *
 * Unrelated roles, conditions, members, etag, and policy metadata are retained
 * byte-for-byte. The result is always policy version 3 because writing a v1
 * representation can silently corrupt conditional bindings.
 */
export function replaceOwnedIamBinding(options: {
  policy: IamPolicy;
  role: string;
  previousCondition?: unknown;
  targetCondition?: unknown;
  previousMembers: ReadonlySet<string>;
  targetMembers: ReadonlySet<string>;
}): IamPolicy {
  const validated = validatedIamPolicy(options.policy);
  const bindings = (validated.bindings ?? []) as IamBinding[];
  const retained: IamBinding[] = [];

  for (const raw of bindings) {
    const binding = structuredClone(raw);
    if (
      binding.role !== options.role ||
      !sameCondition(binding.condition, options.previousCondition)
    ) {
      retained.push(binding);
      continue;
    }
    if (!Array.isArray(binding.members)) {
      throw new Error("IAM binding contains an invalid members collection");
    }
    const members = binding.members.filter((member): member is string => {
      if (typeof member !== "string") {
        throw new Error("IAM binding contains a non-string member");
      }
      return !options.previousMembers.has(member);
    });
    if (members.length > 0) retained.push({ ...binding, members });
  }

  const target = retained.find(
    (binding) =>
      binding.role === options.role &&
      sameCondition(binding.condition, options.targetCondition),
  );
  if (target !== undefined) {
    if (!Array.isArray(target.members)) {
      throw new Error("IAM binding contains an invalid members collection");
    }
    const current = target.members.map((member) => {
      if (typeof member !== "string") {
        throw new Error("IAM binding contains a non-string member");
      }
      return member;
    });
    target.members = [...new Set([...current, ...options.targetMembers])].sort();
  } else if (options.targetMembers.size > 0) {
    retained.push({
      role: options.role,
      members: [...options.targetMembers].sort(),
      ...(options.targetCondition == null ? {} : { condition: options.targetCondition }),
    });
  }

  return validatedIamPolicy({
    ...validated,
    bindings: retained,
    version: 3,
  });
}

File: secure-gateway-studio/backend/src/sgstudio/domain/iam_policy.py
from __future__ import annotations

import json
from copy import deepcopy
from dataclasses import dataclass
from typing import Any

_POLICY_FIELDS = frozenset({"version", "bindings", "auditConfigs", "etag"})
_BINDING_FIELDS = frozenset({"role", "members", "condition"})
_CONDITION_FIELDS = frozenset({"title", "expression", "description", "location"})
_AUDIT_CONFIG_FIELDS = frozenset({"service", "auditLogConfigs"})
_AUDIT_LOG_CONFIG_FIELDS = frozenset({"logType", "exemptedMembers"})
_AUDIT_LOG_TYPES = frozenset({"ADMIN_READ", "DATA_WRITE", "DATA_READ"})


class IamPolicyValidationError(ValueError):
    """A provider IAM policy cannot be safely round-tripped."""


class IamPolicyEtagMissingError(IamPolicyValidationError):
    """A policy about to be written has no usable concurrency token."""


def _assert_exact_fields(
    value: dict[object, object],
    allowed: frozenset[str],
    *,
    context: str,
) -> None:
    fields = set(value)
    if any(not isinstance(field, str) for field in fields) or not fields <= allowed:
        raise IamPolicyValidationError(f"IAM {context} contains an unknown field")


def _nonempty_string(value: object, *, context: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise IamPolicyValidationError(f"IAM {context} must be a nonempty string")
    return value


def _unique_nonempty_strings(
    value: object,
    *,
    context: str,
    require_item: bool,
) -> list[str]:
    if not isinstance(value, list) or (require_item and not value):
        raise IamPolicyValidationError(
            f"IAM {context} must be a{' nonempty' if require_item else ''} list"
        )
    if any(not isinstance(item, str) or not item.strip() for item in value):
        raise IamPolicyValidationError(f"IAM {context} contains an empty or non-string value")
    if len(set(value)) != len(value):
        raise IamPolicyValidationError(f"IAM {context} contains a duplicate value")
    return list(value)


def _validate_condition(value: object) -> dict[str, str]:
    if not isinstance(value, dict):
        raise IamPolicyValidationError("IAM binding condition must be an object")
    _assert_exact_fields(value, _CONDITION_FIELDS, context="binding condition")
    condition: dict[str, str] = {}
    for required in ("title", "expression"):
        condition[required] = _nonempty_string(
            value.get(required),
            context=f"binding condition {required}",
        )
    for optional in ("description", "location"):
        if optional in value:
            raw = value[optional]
            if not isinstance(raw, str):
                raise IamPolicyValidationError(
                    f"IAM binding condition {optional} must be a string"
                )
            condition[optional] = raw
    return condition


def _validate_audit_configs(value: object) -> list[dict[str, Any]]:
    if not isinstance(value, list):
        raise IamPolicyValidationError("IAM policy auditConfigs must be a list")
    configs: list[dict[str, Any]] = []
    services: set[str] = set()
    for raw_config in value:
        if not isinstance(raw_config, dict):
            raise IamPolicyValidationError("IAM audit config must be an object")
        _assert_exact_fields(raw_config, _AUDIT_CONFIG_FIELDS, context="audit config")
        service = _nonempty_string(raw_config.get("service"), context="audit config service")
        if service in services:
            raise IamPolicyValidationError("IAM policy contains duplicate audit services")
        services.add(service)
        config: dict[str, Any] = {"service": service}
        if "auditLogConfigs" in raw_config:
            raw_log_configs = raw_config["auditLogConfigs"]
            if not isinstance(raw_log_configs, list):
                raise IamPolicyValidationError("IAM auditLogConfigs must be a list")
            log_configs: list[dict[str, Any]] = []
            log_types: set[str] = set()
            for raw_log_config in raw_log_configs:
                if not isinstance(raw_log_config, dict):
                    raise IamPolicyValidationError("IAM audit log config must be an object")
                _assert_exact_fields(
                    raw_log_config,
                    _AUDIT_LOG_CONFIG_FIELDS,
                    context="audit log config",
                )
                log_type = _nonempty_string(
                    raw_log_config.get("logType"),
                    context="audit log type",
                )
                if log_type not in _AUDIT_LOG_TYPES or log_type in log_types:
                    raise IamPolicyValidationError("IAM audit log type is invalid or duplicated")
                log_types.add(log_type)
                log_config: dict[str, Any] = {"logType": log_type}
                if "exemptedMembers" in raw_log_config:
                    log_config["exemptedMembers"] = _unique_nonempty_strings(
                        raw_log_config["exemptedMembers"],
                        context="audit exemptedMembers",
                        require_item=False,
                    )
                log_configs.append(log_config)
            config["auditLogConfigs"] = log_configs
        configs.append(config)
    return configs


def validate_iam_policy_v3(
    policy: object,
    *,
    require_etag: bool = False,
) -> dict[str, Any]:
    """Return a defensive copy of a strict Google IAM v3-compatible policy.

    Google may return version 1 (or omit the version) after a v3 read when the
    policy has no conditions. Conditional policies must already identify
    themselves as version 3. Callers set ``version`` to 3 before mutation.
    """

    if not isinstance(policy, dict):
        raise IamPolicyValidationError("IAM policy must be an object")
    _assert_exact_fields(policy, _POLICY_FIELDS, context="policy")

    version = policy.get("version")
    if version is not None and (
        isinstance(version, bool) or not isinstance(version, int) or version not in {0, 1, 3}
    ):
        raise IamPolicyValidationError("IAM policy version is invalid")

    etag = policy.get("etag")
    if require_etag and (not isinstance(etag, str) or not etag.strip()):
        raise IamPolicyEtagMissingError("IAM policy is missing a nonempty etag")
    if etag is not None and not isinstance(etag, str):
        raise IamPolicyValidationError("IAM policy etag must be a string")

    raw_bindings = policy.get("bindings", [])
    if not isinstance(raw_bindings, list):
        raise IamPolicyValidationError("IAM policy bindings must be a list")
    bindings: list[dict[str, Any]] = []
    has_condition = False
    for raw_binding in raw_bindings:
        if not isinstance(raw_binding, dict):
            raise IamPolicyValidationError("IAM policy binding must be an object")
        _assert_exact_fields(raw_binding, _BINDING_FIELDS, context="binding")
        binding: dict[str, Any] = {
            "role": _nonempty_string(raw_binding.get("role"), context="binding role"),
            "members": _unique_nonempty_strings(
                raw_binding.get("members"),
                context="binding members",
                require_item=True,
            ),
        }
        if "condition" in raw_binding:
            binding["condition"] = _validate_condition(raw_binding["condition"])
            has_condition = True
        bindings.append(binding)
    if has_condition and version != 3:
        raise IamPolicyValidationError("IAM policy with conditions must use version 3")

    validated: dict[str, Any] = {}
    if "version" in policy:
        validated["version"] = version
    if "etag" in policy:
        validated["etag"] = etag
    if "bindings" in policy:
        validated["bindings"] = bindings
    if "auditConfigs" in policy:
        validated["auditConfigs"] = _validate_audit_configs(policy["auditConfigs"])
    return validated


@dataclass
class _ParsedBinding:
    binding: dict[str, Any]
    key: str
    members: list[str]


def _parse_bindings(policy: dict[str, Any]) -> list[_ParsedBinding]:
    bindings = policy.get("bindings", [])
    parsed: list[_ParsedBinding] = []
    for raw in bindings:
        binding = deepcopy(raw)
        role = binding["role"]
        members = binding["members"]
        key = json.dumps(
            [role, binding.get("condition")],
            ensure_ascii=False,
            sort_keys=True,
            separators=(",", ":"),
        )
        parsed.append(_ParsedBinding(binding=binding, key=key, members=list(members)))
    return parsed


def _member_groups(bindings: list[_ParsedBinding]) -> dict[str, set[str]]:
    groups: dict[str, set[str]] = {}
    for binding in bindings:
        groups.setdefault(binding.key, set()).update(binding.members)
    return groups


def revert_iam_policy_delta(
    *,
    before_policy: dict[str, Any],
    after_policy: dict[str, Any],
    current_policy: dict[str, Any],
) -> dict[str, Any]:
    """Reverse one run's binding-member delta against a freshly read policy.

    Role/condition/member changes made after Apply by another administrator are
    retained. Only members in ``after - before`` are removed and members in
    ``before - after`` are restored. Policy metadata and the concurrency token
    come from ``current_policy``.
    """

    validated_before = validate_iam_policy_v3(before_policy)
    validated_after = validate_iam_policy_v3(after_policy)
    validated_current = validate_iam_policy_v3(current_policy, require_etag=True)
    before_bindings = _parse_bindings(validated_before)
    after_bindings = _parse_bindings(validated_after)
    reverted = _parse_bindings(validated_current)
    before_groups = _member_groups(before_bindings)
    after_groups = _member_groups(after_bindings)

    for key in sorted(before_groups.keys() | after_groups.keys()):
        before_members = before_groups.get(key, set())
        after_members = after_groups.get(key, set())
        added = after_members - before_members
        removed = before_members - after_members
        if not added and not removed:
            continue

        target = next((entry for entry in reverted if entry.key == key), None)
        for entry in reverted:
            if entry.key == key:
                entry.members = [member for member in entry.members if member not in added]

        if removed:
            if target is None:
                template = next(
                    (entry for entry in (*before_bindings, *after_bindings) if entry.key == key),
                    None,
                )
                if template is None:
                    raise ValueError("IAM delta lacks a binding template")
                target = _ParsedBinding(
                    binding=deepcopy(template.binding),
                    key=key,
                    members=[],
                )
                reverted.append(target)
            target.members = sorted(set(target.members) | removed)

        reverted = [entry for entry in reverted if entry.key != key or entry.members]

    result = deepcopy(validated_current)
    result["bindings"] = [{**entry.binding, "members": entry.members} for entry in reverted]
    result["version"] = 3
    return validate_iam_policy_v3(result, require_etag=True)
